common/wuanalysis/anawu.cpp

```diff
@@ -0,0 +1,429 @@
+    WuScope * parent = nullptr;
+    Owned<IPropertyTree> attrs;
+    WuScopeHashTable scopes;
+    std::vector<WuScope *> inputs;
+    std::vector<WuScope *> outputs;
+};
+
+//-----------------------------------------------------------------------------------------------------------
+class WorkunitAnalyser
+{
+public:
+    WorkunitAnalyser(WuAnalyseOptions & _options);
+
+    void check(const char * scope, IWuActivity & activity);
+    void check(IConstWorkUnit * wu);
+
+protected:
+    void collateWorkunitStats(IConstWorkUnit * workunit, const WuScopeFilter & filter);
+    void limitResultsPerScope();
+    void printWarnings();
+    WuScope * selectFullScope(const char * scope);
+
+protected:
+    CIArrayOf<AActivityRule> rules;
+    CIArrayOf<PerformanceIssue> issues;
+    WuScope root;
+    WuAnalyseOptions & options;
+};
+
+//-----------------------------------------------------------------------------------------------------------
+void WuScopeHashTable::onRemove(void *et)
+{
+    WuScope * elem = reinterpret_cast<WuScope *>(et);
+    elem->Release();
+}
+unsigned WuScopeHashTable::getHashFromElement(const void *et) const
+{
+    const WuScope * elem = reinterpret_cast<const WuScope *>(et);
+    return hashScope(elem->queryName());
+}
+
+unsigned WuScopeHashTable::getHashFromFindParam(const void *fp) const
+{
+    const char * search = reinterpret_cast<const char *>(fp);
+    return hashScope(search);
+}
+
+const void * WuScopeHashTable::getFindParam(const void *et) const
+{
+    const WuScope * elem = reinterpret_cast<const WuScope *>(et);
+    return elem->queryName();
+}
+
+bool WuScopeHashTable::matchesFindParam(const void *et, const void *key, unsigned fphash) const
+{
+    const WuScope * elem = reinterpret_cast<const WuScope *>(et);
+    const char * search = reinterpret_cast<const char *>(key);
+    return streq(elem->queryName(), search);
+}
+
+bool WuScopeHashTable::matchesElement(const void *et, const void *searchET) const
+{
+    const WuScope * elem = reinterpret_cast<const WuScope *>(et);
+    const WuScope * searchElem = reinterpret_cast<const WuScope *>(searchET);
+    return streq(elem->queryName(), searchElem->queryName());
+}
+
+//-----------------------------------------------------------------------------------------------------------
+void WuScope::applyRules(WorkunitAnalyser & analyser)
+{
+    for (auto & cur : scopes)
+    {
+        if (cur.queryScopeType() == SSTactivity)
+            analyser.check(cur.queryName(), cur);
+        cur.applyRules(analyser);
+    }
+}
+
+void WuScope::connectActivities()
+{
+    //Yuk - scopes can be added to while they are being iterated, so need to create a list first
+    CICopyArrayOf<WuScope> toWalk;
+    for (auto & cur : scopes)
+        toWalk.append(cur);
+
+    ForEachItemIn(i, toWalk)
+    {
+        WuScope & cur = toWalk.item(i);
+
+        if (cur.queryScopeType() == SSTedge)
+        {
+            const char * tmp = attrs->queryProp("@IdSource");
```

Comment:
Looks like old debugging code

Reply:
Yes, this slipped in by mistake...removed.